Use `Write.UTxO` in `ErrBalanceTxInsufficientCollateralError`.

jonathanknowles opened the pull request:
## Issue

ADP-3184

## Description

This PR adjusts `ErrBalanceTxInsufficientCollateralError` to use `Write.UTxO` instead of the wallet primitive `UTxO` type.

jonathanknowles committed:
Add `era` parameter to `ErrBalanceTxInsufficientCollateralError`.

jonathanknowles committed:
Add `era` parameter to `coinSelectionErrorToBalanceTxError`.

jonathanknowles committed:
Derive standalone instances for `ErrBalanceTxInsufficientCollateralError`.

jonathanknowles committed:
Change `largestCombinationAvailable` to use `Write.UTxO` type.

Note: this commit breaks the "pingPong_2" golden test for `balanceTransaction`.

The next commit will fix this test.

jonathanknowles commented on lib/wallet-e2e/src/Cardano/Wallet/Spec/Effect/Trace.hs:
```diff
@@ -35,6 +35,7 @@ import Path.IO
 import Prelude hiding
     ( modify
     , runState
+    , trace
```
@Anviking 
We actually include a similar change in https://github.com/cardano-foundation/cardano-wallet/pull/4181 (https://github.com/cardano-foundation/cardano-wallet/pull/4181/commits/8e13bd4096d9d8c9ba1ddf69452f8e449df0f82c).

jonathanknowles committed:
Add type signature to `tipSlotStartTime`.

jonathanknowles committed:
Add type signature to `trWorker`.

jonathanknowles committed:
Add type signature to `toMintTxOut`.

jonathanknowles committed:
Add type signature to `mintingOuts`.

jonathanknowles committed:
Add type signature to `coalesceTokensPerAddr`.

jonathanknowles committed:
Add type signature to `benchEstimateTxFee`.

jonathanknowles committed:
Add type signature to `forAllRecentEras'`.

jonathanknowles committed:
Use concrete monadic IO type in `BalanceSpec`.

jonathanknowles committed:
Fix golden tests for `balanceTransaction`.

Anviking commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -3152,6 +3153,7 @@ constructSharedTransaction
     withWorkerCtx api wid liftE liftE $ \wrk -> do
         let db = wrk ^. dbLayer
             netLayer = wrk ^. networkLayer
+            trWorker :: Tracer IO W.WalletLog
```
Are these required now?

jonathanknowles replied:
> Are these required now?

Yes, it seems so.

I added type annotations/signatures one by one until I could get the whole codebase to compile.

jonathanknowles commented on lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:
```diff
@@ -2122,6 +2122,10 @@ hasTotalCollateral (CardanoApi.Tx (CardanoApi.TxBody content) _) =
         CardanoApi.TxTotalCollateral _ _ -> True
 
+-- | A type-constrained variant of 'it'.
+itIO :: String -> IO () -> Spec
```
Perhaps we could add a TODO here and link it to a ticket that reminds us to remove this when possible?

jonathanknowles replied:
Fixed in 868dbcdb21085da7814b2c936a8d90cafa60e817.

jonathanknowles committed:
Add reminder to consider removing the helper function `itIO`.

In response to review feedback:

https://github.com/cardano-foundation/cardano-wallet/pull/4178/files#r1387690723

Anviking commented on lib/wallet/test/data/balanceTx/pingPong_2/golden:
```diff
@@ -39,26 +39,26 @@
- 2.950000,ErrBalanceTxInsufficientCollateral (ErrBalanceTxInsufficientCollateralError {largestCombinationAvailable = UTxO {unUTxO = fromList [(TxIn {inputId = Hash "00000000000000000000000000000000", inputIx = 0},TxOut {address = Address "`\177\229\224\251t\200l\128\USdhA\224|\219B\223\139\130\239<\228\229|\181A.w", tokens = TokenBundle {coin = Coin 2950000, tokens = TokenMap (fromList [])}})]},  ...
- 3.000000,ErrBalanceTxInsufficientCollateral (ErrBalanceTxInsufficientCollateralError {largestCombinationAvailable = UTxO {unUTxO = fromList [(TxIn {inputId = Hash "00000000000000000000000000000000", inputIx = 0},TxOut {address = Address "`\177\229\224\251t\200l\128\USdhA\224|\219B\223\139\130\239<\228\229|\181A.w", tokens = TokenBundle {coin = Coin 3000000, tokens = TokenMap (fromList [])}})]},  ...
+ 2.050000,ErrBalanceTxInsufficientCollateral (ErrBalanceTxInsufficientCollateralError {largestCombinationAvailable = UTxO (fromList [(TxIn (TxId {unTxId = SafeHash "3030303030303030303030303030303030303030303030303030303030303030"}) (TxIx 0),(Addr Testnet (KeyHashObj (KeyHash "b1e5e0fb74c86c801f646841e07cdb42df8b82ef3ce4e57cb5412e77")) StakeRefNull,MaryValue 2050000 (MultiAsset (fromList [])),NoD ...
+ 2.100000,ErrBalanceTxInsufficientCollateral (ErrBalanceTxInsufficientCollateralError {largestCombinationAvailable = UTxO (fromList [(TxIn (TxId {unTxId = SafeHash "3030303030303030303030303030303030303030303030303030303030303030"}) (TxIx 0),(Addr Testnet (KeyHashObj (KeyHash "b1e5e0fb74c86c801f646841e07cdb42df8b82ef3ce4e57cb5412e77")) StakeRefNull,MaryValue 2100000 (MultiAsset (fromList [])),NoD ...
```
👍

Anviking reviewed (approved):
The fragile 8.10 type inference with `IsRecentEra` in its current state might cause confusion, but seems fine to me to go with for now and hope https://cardanofoundation.atlassian.net/browse/ADP-2353 or something similar will fix 👍
